ee/clickhouse/queries/cohort_query.py

```diff
@@ -0,0 +1,476 @@
+        return list(set(self._events))
+
+
+class CohortQuery(EnterpriseEventQuery):
+
+    BEHAVIOR_QUERY_ALIAS = "behavior_query"
+    _cohort_optimizer: CohortOptimizer
+
+    def __init__(
+        self,
+        filter: Union[Filter, PathFilter, RetentionFilter, StickinessFilter, SessionRecordingsFilter],
+        team: Team,
+        round_interval=False,
+        should_join_distinct_ids=False,
+        should_join_persons=False,
+        # Extra events/person table columns to fetch since parent query needs them
+        extra_fields: List[ColumnName] = [],
+        extra_event_properties: List[PropertyName] = [],
+        extra_person_fields: List[ColumnName] = [],
+        override_aggregate_users_by_distinct_id: Optional[bool] = None,
+        **kwargs,
+    ) -> None:
+        self._cohort_optimizer = CohortOptimizer(filter, team.pk)
+        super().__init__(
+            filter=filter,
+            team=team,
+            round_interval=round_interval,
+            should_join_distinct_ids=should_join_distinct_ids,
+            should_join_persons=should_join_persons,
+            extra_fields=extra_fields,
+            extra_event_properties=extra_event_properties,
+            extra_person_fields=extra_person_fields,
+            override_aggregate_users_by_distinct_id=override_aggregate_users_by_distinct_id,
+            **kwargs,
+        )
+
+    def get_query(self) -> Tuple[str, Dict[str, Any]]:
+
+        behavior_subquery, behavior_subquery_params = self._get_behavior_subquery()
+        self.params.update(behavior_subquery_params)
+        conditions, condition_params = self._get_conditions()
+        self.params.update(condition_params)
+
+        person_query, person_params = self._get_person_query()
+        self.params.update(person_params)
+
+        final_query = f"""
+        SELECT person_id FROM
+        ({behavior_subquery}) {self.BEHAVIOR_QUERY_ALIAS}
+        {person_query}
+        WHERE team_id = %(team_id)s
+        {conditions}
+        """
+
+        return final_query, self.params
+
+    def _get_behavior_subquery(self) -> Tuple[str, Dict[str, Any]]:
+        """
+        Get the subquery for the cohort query.
+        """
+
+        _fields = [f"{self.DISTINCT_ID_TABLE_ALIAS}.person_id as person_id"]
+
+        date_fields = self._get_date_fields()
+        event_in_period_fields = self._get_event_in_period_fields()
+        first_time_activity_fields = self._get_first_time_fields()
+        sequence_fields = self._get_sequence_fields()
+
+        _fields.extend(date_fields)
+        _fields.extend(event_in_period_fields)
+        _fields.extend(first_time_activity_fields)
+        _fields.extend(sequence_fields)
+
+        events = self._cohort_optimizer.get_events_to_query()
+
+        query = f"""
+        SELECT {", ".join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}
+        {self._get_distinct_id_query()}
+        WHERE team_id = %(team_id)s
+        AND event IN %(events)s
```

Comment:
We always want a date filter in here! If each field has its own date range, this should be the max of all of them.

Reply:
good point